test/server/config_validation/xds_fuzz.cc

```diff
@@ -0,0 +1,244 @@
+    ads_cluster->set_name("ads_cluster");
+  });
+  setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+  HttpIntegrationTest::initialize();
+  if (xds_stream_ == nullptr) {
+    createXdsConnection();
+    AssertionResult result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    xds_stream_->startGrpcStream();
+  }
+}
+
+void XdsFuzzTest::close() {
+  cleanUpXdsConnection();
+  test_server_.reset();
+  fake_upstreams_.clear();
+}
+
+/**
+ * remove a listener from the list of listeners if it exists
+ * @param the listener number to be removed
+ * @return the listener as an optional so that it can be used in a delta request
+ */
+absl::optional<std::string> XdsFuzzTest::removeListener(uint32_t listener_num) {
+  std::string match = absl::StrCat("listener_", listener_num % num_listeners_);
+
+  for (auto it = listeners_.begin(); it != listeners_.end(); ++it) {
+    if (it->name() == match) {
+      listeners_.erase(it);
+      return match;
+    }
+  }
+  return {};
+}
+
+/**
+ * remove a route from the list of routes if it exists
+ * @param the route number to be removed
+ * @return the route as an optional so that it can be used in a delta request
+ */
+absl::optional<std::string> XdsFuzzTest::removeRoute(uint32_t route_num) {
+  std::string match = absl::StrCat("route_config_", route_num % num_routes_);
+  for (auto it = routes_.begin(); it != routes_.end(); ++it) {
+    if (it->name() == match) {
+      routes_.erase(it);
+      return match;
+    }
+  }
+  return {};
+}
+
+/**
+ * run the sequence of actions defined in the fuzzed protobuf
+ */
+void XdsFuzzTest::replay() {
+  initialize();
+
+  // set up cluster
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "", {}, {}, {}, true));
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(Config::TypeUrl::get().Cluster,
+                                                             {buildCluster("cluster_0")},
+                                                             {buildCluster("cluster_0")}, {}, "0");
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().ClusterLoadAssignment, "",
+                                      {"cluster_0"}, {"cluster_0"}, {}));
+  sendDiscoveryResponse<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+      Config::TypeUrl::get().ClusterLoadAssignment, {buildClusterLoadAssignment("cluster_0")},
+      {buildClusterLoadAssignment("cluster_0")}, {}, "0");
+
+  for (const auto& action : actions_) {
+    switch (action.action_selector_case()) {
+    case test::server::config_validation::Action::kAddListener: {
+      removeListener(action.add_listener().listener_num());
+      auto listener =
+          buildListener(action.add_listener().listener_num(), action.add_listener().route_num());
+      listeners_.push_back(listener);
+
+      updateListener(listeners_, {listener}, {});
+      // TODO(samflattery): compareDiscoveryResponse to check ACK/NACK?
```

Comment:
Is the plan to also compare the actual listeners on the server with those we think should be there (i.e. from `listeners_`)?

Reply:
yep, i'm also writing a class that stores an abstract representation of what Envoy should be doing, such as storing each listener and its {warming, draining, active} state, which I'm then comparing to what is actually on the server.  This will probably be in another PR as it's not ready yet

Reply:
Sure, whatever works best here. Happy to launch-and-iterate.